### DISCUSSION TITLE
Automatic commenting on a commit with results from a script

### DISCUSSION THREAD

**Discussion by sayakpaul** [community/community#26914] [Actions]
<p>Hi folks!</p>
<p>I am trying to create a comment on a commit automatically by using results from my runner program. Let’s consider <a href="https://github.com/sayakpaul/wine/runs/979690608?check_suite_focus=true" rel="nofollow noopener">this run</a>. I want to be able to take the highlighted URL and include that in the comment:</p>
<p><img src="https://user-images.githubusercontent.com/22957388/181100599-e4cd1371-7dfc-448c-8361-7d094d4dc38b.png" alt="Screen Shot 2020-08-28 at 9.02.08 PM" data-base62-sha1="rnJ8MqXMeJ8m6Ws05xNjpDhCEFI" width="651" height="296"></p>
<p>Programmatically, I am using <a href="http://wandb.com" rel="nofollow noopener">Weights and Biases</a> to keep track of my ML models. So, once, I run <code>wandb.init(project="wandb-github-actions", entity="sayakpaul")</code> from my Python program I am able to access that URL via <code>wandb.run.url</code> within the program.</p>

**Comment by NicoNBW** [comment]
<p>Thanks for your feedback.<br>
From your train.py file(<a href="https://github.com/sayakpaul/wine/blob/3f59c17a8ddd6504211ec35d40ca716b9111321e/train.py#L51" rel="nofollow noopener">Line 51~55</a>), did you want to write the <code>run_url</code> into the <code>metrics.txt</code> file, and then allow this data to be accessed by subsequent Actions?<br>
If so, you could add reading the <code>metrics.txt</code> file step to check if this <code>run_url</code> is correct and then use git command to push this commit.<br>
You can also add git command directly in your train.py file, <a href="https://github.com/TestWorkflowsRML/AddTwoNumbers/blob/c9d47c1723ecb753e1af17536ae8da472ec575e7/create-update-file.py" rel="nofollow noopener">such as</a>:</p>
<pre><code class="lang-auto">print("\n======================= Push changes to remote =======================")
import subprocess
print("Set git config...")
subprocess.Popen(["git", "config", "--global", "user.name", "github-actions[bot]"])
subprocess.Popen(["git", "config", "--global", "user.email", "41898282+github-actions[bot]@users.noreply.github.com"])

print("Push to remote...")
subprocess.Popen(["git", "add", "-A"])
subprocess.Popen(["git", "commit", "-m", "Your run url"])
subprocess.Popen(["git", "push"])
</code></pre>
<p>Hope it helps and if I misunderstood you, please correct me. Thanks</p>

**Comment by sayakpaul** [comment]
<p>Hi <a class="mention" href="/niconbw">@niconbw</a> yes, you understood the requirement correctly. I wanted to understand where the comment would go?</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a></p>
<aside class="quote no-group" data-username="sayakpaul" data-post="3" data-topic="129529">
<div class="title">
<div class="quote-controls"></div>
 sayakpaul:</div>
<blockquote>
<p>I wanted to understand where the comment would go?</p>
</blockquote>
</aside>
<p>Did you mean to know where the <code>metrics.txt</code> file is generated? If so, the file will be saved into the path specified by the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">default environment variable</a> <code>github.workspace</code>. Then you can directly access this default working directory.</p>

**Comment by sayakpaul** [comment]
<p><span class="mention">@niconbwI</span> was meaning where the <strong>comment</strong> would go? Because my requirement is to be able to log the run URL as a comment upon each new commit.</p>

**Comment by NicoNBW** [comment, marked as answer]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
I can’t run your workflow because there is no wandb environment. Generally speaking, if you can get the value of the <code>run_url</code> in your train.py file, you can use this <code>run_url</code> as the comment when submitting the commit.<br>
<a href="https://github.com/NicoNBW/testShowUrl/blob/bc5b41f2a2c00a50ad9d78a3f54ed706b0a3e781/create-update-file.py#L14" rel="nofollow noopener">A simple example</a> :</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/69940968e5bc400cc0281d0a27b347201d901d22.png" data-download-href="/uploads/short-url/f3ZiHGd4HZJ3wmEDxKpu2kqgbZg.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46212442/181100608-4e170514-39b6-4b8a-9a6e-a4285741ca59.png" alt="image" data-base62-sha1="f3ZiHGd4HZJ3wmEDxKpu2kqgbZg" width="690" height="327" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/69940968e5bc400cc0281d0a27b347201d901d22_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">852×405 20 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Result:<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/55a1330502854481353a7e81925f5728868e4078.png" data-download-href="/uploads/short-url/cdvX4TfpyHebLM7EVeHslJmMtiE.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46212442/181100609-07e09fda-ce3e-4dd0-a344-b70b74aa59fc.png" alt="image" data-base62-sha1="cdvX4TfpyHebLM7EVeHslJmMtiE" width="690" height="130" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/5/55a1330502854481353a7e81925f5728868e4078_2_690x130.png, https://aws1.discourse-cdn.com/github/optimized/2X/5/55a1330502854481353a7e81925f5728868e4078_2_1035x195.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/5/55a1330502854481353a7e81925f5728868e4078_2_1380x260.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/55a1330502854481353a7e81925f5728868e4078_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1554×293 34.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If it still doesn’t meet your needs, some screenshots for describing your request will be appreciated.<br>
By the way, there is an existing <a href="https://github.com/machine-learning-apps/wandb-action" rel="nofollow noopener">wandb action</a>. You could check it. Hope it helps.</p>

**Comment by sayakpaul** [comment]
<p><span class="mention">@niconbwappreciate</span> your help here. I am aware of that repository but I found it kind of complicated to get started with. Hence I preferred something simpler.</p>
<p>Your example does meet the requirement. Do you think it would be better to run this action on a PR and have the run URL as a comment?</p>
<p>Also, if you set your <code>wandb</code> API Key as a GitHub secret my workflow should be able to fetch that and utilize it.</p>
<p>Again, thank you so much for being so helpful :slight_smile:</p>

**Comment by NicoNBW** [comment]
<p>Thanks for your update.<br>
The action above will save a csv file with more information. You can decide whether to use it according to your needs.</p>

**Comment by sayakpaul** [comment]
<p>Thanks, <a class="mention" href="/niconbw">@niconbw</a>.  So, if I were to extend this use-case to support commenting capabilities upon a new PR what would be the solutions?</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
Not sure but you could try to use <a href="https://docs.github.com/en/rest/reference/pulls#create-a-pull-request" rel="nofollow noopener">Create Pull Request REST API</a> in your<code> train.py</code> file.<br>
More about <a href="https://rapidapi.com/blog/how-to-use-an-api-with-python/" rel="nofollow noopener">How to use an API with Python</a>. Hope it helps.</p>

**Comment by hamelsmu** [comment]
<p><a class="mention" href="/sayakpaul">@sayakpaul</a> have you seen the weights and biases GitHub Action?  <a href="https://github.com/marketplace/actions/get-runs-from-weights-biases" rel="nofollow noopener">https://github.com/marketplace/actions/get-runs-from-weights-biases</a></p>

**Comment by hamelsmu** [comment]
<p>To comment on a PR, its best to use this Action <a href="https://github.com/actions/github-script" rel="nofollow noopener">https://github.com/actions/github-script</a></p>

**Comment by sayakpaul** [comment]
<p><a class="mention" href="/hamelsmu">@hamelsmu</a> thanks for your response. I am aware of that action but my purpose with the action I mentioned in this thread is a bit more reduced. I want to be able to (upon a new commit) -</p>
<ul>
<li>Train a toy ML model</li>
<li>Log experiments to <code>wandb</code>
</li>
<li>Generate artifacts file and provide them as an action artifact download</li>
<li>Cache pip dependencies</li>
</ul>
<p>It might seem very trivial to you but I am doing this for self-learning :smiley:</p>
<p>Then upon a PR, I want to be able to fetch the latest run URL and comment that out on the PR. I have been able to realize these so far - <a href="https://github.com/sayakpaul/wine/blob/a49b0d8386c6ef926bca45fd1cd2e8cd474a52b0/.github/workflows" rel="nofollow noopener">https://github.com/sayakpaul/wine/blob/a49b0d8386c6ef926bca45fd1cd2e8cd474a52b0/.github/workflows</a> with a caveat that my action is picking up on an older run instead of the latest one.</p>
